File: cpp/1647.minimum-deletions-to-make-character-frequencies-unique.cpp
class Solution {
public:
    int minDeletions(string s) {
        int n = 26;

        vector<int> freq(n, 0);
        for (char ch : s) {
            freq[ch - 'a']++;
        }
        sort(freq.begin(), freq.end(), greater<int>());
        for (auto fr : freq) { cout << fr << " ";}
        cout << endl;

        int deletions = 0;
        // && freq[i] !=0 && freq[i + 1] != 0;
        for (int i = 0; i < n - 1; i++) {
            if (freq[i] == 0 && freq[i] <= freq[i + 1]) {
                auto diff = freq[i + 1] - freq[i];
                deletions += diff;
                freq[i + 1] -= diff;
            } else if (freq[i] <= freq[i + 1]) {
                auto diff = 1 + freq[i + 1] - freq[i];
                deletions += diff;
                freq[i + 1] -= diff;
            }
        }

        for (auto fr : freq) { cout << fr << " ";}

        return deletions;

    }
};

File: golang/1647.minimum-deletions-to-make-character-frequencies-unique.go
func minDeletions(s string) int {
    n := 26

    frq := make([]int, n)
    for _, ch := range(s) {
        frq[ch - 'a']++
    }

    sort.Sort(sort.Reverse(sort.IntSlice(frq)))

    deletions := 0
    for i := 0; i < n - 1 && frq[i + 1] != 0; i++ {
        if frq[i] == 0 && frq[i + 1] >= 0 {
            deletions += frq[i + 1]
            frq[i + 1] = 0
        } else if frq[i] <= frq[i + 1] {
            diff := 1 + frq[i + 1] - frq[i]
            deletions += diff
            frq[i + 1] -= diff
        }
    }

    return deletions
}